File: app/services/scenarios.py
"""Per-user parallel thesis "scenarios" (ADR-026) - shared resolution logic
used by every write action that needs to know "which user's thesis am I
acting on."

Design, confirmed with the user: actions implicitly apply to the caller's
own scenario. There is no explicit scenario_id in request URLs - "amend
the thesis" always means "amend my thesis." A scenario is created exactly
once, when a user first writes a thesis on a company (via create_company
for a brand-new company, or start_scenario for a company that already
exists under someone else's thesis). Every other write (amend, health
check, decision) requires an existing scenario and 404s if the caller
hasn't started one yet, rather than silently creating one on the wrong verb.
"""
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import Company, ThesisScenario


class NotFoundError(Exception):
    pass


class ScenarioNotFoundError(Exception):
    """Raised when the caller has no scenario on this company yet - distinct
    from NotFoundError (company doesn't exist at all) so routers can return
    a message pointing at "start your own thesis" instead of a bare 404."""


def get_company_or_404(db: Session, company_id: str) -> Company:
    company = db.get(Company, company_id)
    if company is None:
        raise NotFoundError(f"company '{company_id}' not found")
    return company


def get_my_scenario(db: Session, company_id: str, owner: str) -> ThesisScenario:
    get_company_or_404(db, company_id)
    scenario = db.scalar(
        select(ThesisScenario).where(ThesisScenario.company_id == company_id, ThesisScenario.owner == owner)
    )
    if scenario is None:
        raise ScenarioNotFoundError(
            f"'{owner}' has no thesis on company '{company_id}' yet - start one first"
        )
    return scenario


def get_scenario_optional(db: Session, company_id: str, owner: str) -> ThesisScenario | None:
    return db.scalar(
        select(ThesisScenario).where(ThesisScenario.company_id == company_id, ThesisScenario.owner == owner)
    )


def list_scenarios(db: Session, company_id: str) -> list[ThesisScenario]:
    return db.scalars(
        select(ThesisScenario).where(ThesisScenario.company_id == company_id).order_by(ThesisScenario.created_at)
    ).all()

File: web/src/lib/server/services/scenarios.ts
// Ports app/services/scenarios.py
import { and, asc, eq } from 'drizzle-orm';
import { db } from '../db';
import { companies, thesisScenarios } from '../db/schema';

export class NotFoundError extends Error {}
export class ScenarioNotFoundError extends Error {}

export async function getCompanyOr404(companyId: string) {
	const [company] = await db.select().from(companies).where(eq(companies.companyId, companyId)).limit(1);
	if (!company) throw new NotFoundError(`company '${companyId}' not found`);
	return company;
}

export async function getMyScenario(companyId: string, owner: string) {
	await getCompanyOr404(companyId);
	const [scenario] = await db
		.select()
		.from(thesisScenarios)
		.where(and(eq(thesisScenarios.companyId, companyId), eq(thesisScenarios.owner, owner)))
		.limit(1);
	if (!scenario) {
		throw new ScenarioNotFoundError(`'${owner}' has no thesis on company '${companyId}' yet - start one first`);
	}
	return scenario;
}

export async function getScenarioOptional(companyId: string, owner: string) {
	const [scenario] = await db
		.select()
		.from(thesisScenarios)
		.where(and(eq(thesisScenarios.companyId, companyId), eq(thesisScenarios.owner, owner)))
		.limit(1);
	return scenario ?? null;
}

export async function listScenarios(companyId: string) {
	return db
		.select()
		.from(thesisScenarios)
		.where(eq(thesisScenarios.companyId, companyId))
		.orderBy(asc(thesisScenarios.createdAt));
}
